### PR TITLE
mgr/cli lightweight get method api call

### PR DESCRIPTION
**Pull request by Waadkh7**
Adding new module to call get method with threads and different number of calls.

Fixes: https://tracker.ceph.com/issues/50311
Signed-off-by: Waad AlKhoury <walkhour@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by Waadkh7** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
is this module supposed to be used by end user or just be developers? and why not use [cProfile](https://docs.python.org/3/library/profile.html) ?

**Comment by sebastian-philipp** [comment]
> is this module supposed to be used by end user or just be developers?

.. or Teuthology?

**Comment by pereman2** [comment]
@tchaikov For now it's supposed to be a developer tool, we need it to stress test the mgr `get` with help of the injector (https://github.com/ceph/ceph/pull/41260). 
Well, cProfile looks like a bit overkill for a simple function, let me know if you think we could benefit from it.

**Comment by sebastian-philipp** [comment]
why not add this to mgr/selftest for now? Injecting data and then verifying the performace in a teuthology test sounds like a sane idea to me

**Comment by tchaikov** [comment]
> @tchaikov For now it's supposed to be a developer tool, we need it to stress test the mgr `get` with help of the injector (#41260). The reason for this at the end is to test the addition of caching.
> Well, cProfile looks like a bit overkill for a simple function, let me know if you think we could benefit from it.

@Waadkh7 i think we should leverage existing tools if it's able to address our needs. i just created #41591 which allows us to run code in the selftest module. in other words, we are able to run existing tools using it on mgr.

**Comment by epuertat** [comment]
@tchaikov this PR is a helper for the mgr API Caching activities, but also for developers and users to debug and interact with the mgr aPI (basically another iteration of what I started with https://github.com/ceph/ceph/pull/34840).

We want a tool that gives us the following:

1. List all ceph-mgr API methods and their syntax and documentation, same as with any other Ceph CLI command.
2. Invoke a single API method from the CLI (mostly for devels, users and operators to functionally debug the mgr API).
3. Perform threaded benchmarking against the mgr API methods.

The repl approach is ok but it doesn't cover 1. and 3. (I guess 2. could be achieved by feeding a Python script into the interpreter), and it's more devel-friendly that user-friendly. cProfile is IMHO an overkill if you just want to measure the latency of a high-level method, and doesn't solve the concurrency benchmarking. And same for timeit. Yes, all those tools can solve part of the issue, but not all.

File: src/pybind/mgr/cli_api/module.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            # would presumably have some more appropriate frequency.
+            sleep_interval = self.mgr_tick_period
+            self.log.error('Sleeping for %d seconds', sleep_interval)
+            self.event.wait(sleep_interval)
+            self.event.clear()
+
+    def get_method(self, arg: str):
+        def wrapper():
+            return self.get(arg)
+        return wrapper
```
Comment (epuertat) [inline src/pybind/mgr/cli_api/module.py:65]: There's a Pythonic way to do this: [functools.partial](https://docs.python.org/3/library/functools.html#functools.partial)
```diff
- benchmark_runner.start(self.get_method(arg))
+ benchmark_runner.start(functools.partial(self.get, self, arg)) # you need to bind both self and the first arg
```

**Review by epuertat — COMMENTED** [review]
I know there's a follow up PR for this, but I promised to review it and that's what I'm doing ;-)

I just tried a few times and this is what I get:
```py
>>> ceph mgr api get --help
# No help!
```

```py
>>> ceph mgr api get
null
1.0967254638671875e-05
```
BTW this shouldn't happen. @tchaikov I have the feeling that the CLICommand parser is not properly distinguishing mandatory and optional args (there might be an issue with the default index).

```py
>>> echo $? 
0
# a wrong command returns 0 (ok) instead of an error code and an error message?

>>> ceph mgr api get invented_map
null
2.1219253540039062e-05
>>> echo $? 
0
# same here 
```

And lastly, and more importantly, have you verified that the list you return is correct? I decided to dump the `shared_list` and this is what I get:
```py
# ceph mgr api benchmark get osd_map 100 10
2021-06-10T17:36:46.250+0000 7fed48b96700 -1 WARNING: all dangerous and experimental features are enabled.
2021-06-10T17:36:46.263+0000 7fed4359e700 -1 WARNING: all dangerous and experimental features are enabled.
{"data": [0.00048470497131347656, 0.0005557537078857422, 0.00048732757568359375, 0.0005564689636230469, 0.0005829334259033203, 0.00038504600524902344, 0.0009906291961669922, 0.0005061626434326172, 0.00049591064453125, 0.0003733634948730469, 0.0003802776336669922, 0.0005617141723632812, 0.0006070137023925781, 0.0005869865417480469, 0.0006308555603027344, 0.0005338191986083984, 0.0007674694061279297, 0.0005700588226318359, 0.000408172607421875, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "min": 0, "max": 0.0009906291961669922, "avg": 0.00010464668273925782}
```

I see lots of zeros there that look suspicious...

File: src/pybind/mgr/cli_api/module.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+class ThreadedBenchmarkRunner:
+    def __init__(self,number_of_total_calls,number_of_parellel_calls):
+        self.lock = threading.RLock()
+        self.number_of_parellel_calls = number_of_parellel_calls
+        self.number_of_total_calls = number_of_total_calls
+        self.number_of_iter = int(number_of_total_calls / number_of_parellel_calls)
+        # We use a shared list object to store the times in every thread
+        self.shared_list = [0] * self.number_of_total_calls
```
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/cli_api/module.py:74]: I think there might be a bug in the way this list is handled by the runners... I got a lot of zeroes when benchmarking this with values 100 and 10.

File: src/pybind/mgr/cli_api/tests/test_cliapi.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        c = ThreadedBenchmarkRunner(10,10)
+        self.assertFalse(c.timer({},0))
+
+    def test_calls(self):
+        number_of_total_calls = 0 
+        c = ThreadedBenchmarkRunner(number_of_total_calls,10)
+        self.assertFalse(number_of_total_calls,0)
+
+    def test_calls(self):
+        number_of_parellel_calls = 0
```
Comment (alfonsomthd) [inline src/pybind/mgr/cli_api/tests/test_cliapi.py:21]:
```suggestion
        number_of_parallel_calls = 0 
```

**Review by alfonsomthd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cli_api/module.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        benchmark_runner = ThreadedBenchmarkRunner(number_of_total_calls,
+                                                    number_of_parellel_calls)
+        benchmark_runner.start(partial(self.get,arg))
+        benchmark_runner.join()
+        stats = benchmark_runner.get_stats()
+        return HandleCommandResult(0, json.dumps(stats), "")
+
+class ThreadedBenchmarkRunner:
+    def __init__(self,number_of_total_calls,number_of_parellel_calls):
+        self.number_of_parellel_calls = number_of_parellel_calls
```
Comment (alfonsomthd) [inline src/pybind/mgr/cli_api/module.py:46]:
```suggestion
        self.number_of_parallel_calls = number_of_parallel_calls
```

File: src/pybind/mgr/cli_api/module.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    NATIVE_OPTIONS = [
+        'mgr_tick_period',
+    ]
+
+    @CLICommand('mgr api get')
+    def api_get(self, arg: str):
+        '''
+        Called by the plugin to fetch named cluster-wide objects from ceph-mgr.
+        :param str data_name: Valid things to fetch is osd_map
```
Comment (alfonsomthd) [inline src/pybind/mgr/cli_api/module.py:23]: I undertsan from the comment that only `osd_map` is allowed but I was able to retrieve the `service_map`:
```
[root@ceph2 ceph]# ceph mgr api get service_map
2021-06-17T11:52:43.333+0000 7fe44c88c700 -1 WARNING: all dangerous and experimental features are enabled.
2021-06-17T11:52:43.345+0000 7fe44c88c700 -1 WARNING: all dangerous and experimental features are enabled.
{"epoch": 6, "services": {"rgw": {"daemons": {"summary": "", "4402": {"start_epoch": 2, "gid": 4402, "addr": "172.20.0.2:0/2769796477", "metadata": {"arch": "x86_64", "ceph_release": "quincy", "ceph_version": "ceph version 17.0.0-5283-g3c870731 (3c870731a740807244afd63ae233e9f5a542a89e) quincy (dev)", "ceph_version_short": "17.0.0-5283-g3c870731", "cpu": "Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz", "distro": "centos", "distro_description": "CentOS Linux 8", "distro_version": "8", "frontend_config#0": "beast port=8000", "frontend_type#0": "beast", "hostname": "ceph2", "id": "8000", "kernel_description": "#1 SMP Thu Jun 10 14:19:48 UTC 2021", "kernel_version": "5.12.10-200.fc33.x86_64", "mem_swap_kb": "4194300", "mem_total_kb": "32236628", "num_handles": "1", "os": "Linux", "pid": "5321", "zone_id": "8864de6d-c72d-4963-ae94-fdc10b55a7e1", "zone_name": "default", "zonegroup_id": "3267d0cf-4401-4c16-bbd7-758e5544efad", "zonegroup_name": "default"}, "task_status": {}, "start_stamp": "2021-06-17T11:11:31.920586+0000"}}}}, "modified": "2021-06-17T11:50:16.507045+0000"}
3.719329833984375e-05
```

File: src/pybind/mgr/tox.ini [hunk 6]
```diff
@@ -47,6 +47,7 @@ commands =
         insights/ \
         pg_autoscaler/ \
         progress/ \
+        cli_api/ \
```
Comment (alfonsomthd) [inline src/pybind/mgr/tox.ini:50]: You also have to add `cli_api` to the other checks (flake8, mypy, ...) and fix the errors:
```
diff --git a/src/pybind/mgr/tox.ini b/src/pybind/mgr/tox.ini
index 301bc9569de..7d58b6e9d4b 100644
--- a/src/pybind/mgr/tox.ini
+++ b/src/pybind/mgr/tox.ini
@@ -42,12 +42,12 @@ commands =
         mgr_util.py \
         tests/ \
         cephadm/ \
+        cli_api/ \
         mds_autoscaler/ \
         orchestrator/ \
         insights/ \
         pg_autoscaler/ \
         progress/ \
-        cli_api/ \
         snap_schedule}
 
 [testenv:mypy]
@@ -65,6 +65,7 @@ commands =
            -m alerts \
            -m balancer \
            -m cephadm \
+           -m cli_api \
            -m crash \
            -m dashboard \
            -m devicehealth \
@@ -105,6 +106,7 @@ modules =
     alerts
     balancer
     cephadm
+    cli_api
     crash
     devicehealth
     diskprediction_local
@@ -127,6 +129,7 @@ modules =
     alerts
     balancer
     cephadm
+    cli_api
     crash
     devicehealth
     diskprediction_local
```

Errors found for mypy (also check flake8):
```
cli_api/__init__.py:7: note: In module imported here:
cli_api/module.py: note: In member "__init__" of class "ThreadedBenchmarkRunner":
cli_api/module.py:53: error: Need type annotation for 'jobs'
cli_api/module.py: note: In member "timer" of class "ThreadedBenchmarkRunner":
cli_api/module.py:91: error: Incompatible types in assignment (expression has type "float", variable has type "int")
```

**Review by alfonsomthd — CHANGES_REQUESTED** [review]

**Review by alfonsomthd — APPROVED** [review]

File: src/pybind/mgr/cli_api/module.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        self._time = 0.0
+        self._self_time = []
+        self._lock = threading.Lock()
+
+    def start(self, fn):
+        if(self._number_of_total_calls and self._number_of_parallel_calls):
+            for threadID in range(self._number_of_parallel_calls):
+                new_thread = threading.Thread(target=ThreadedBenchmarkRunner.timer, args=(self, self._jobs, fn,))
+                self._threads.append(new_thread)
+            for jobsID in range(self._number_of_total_calls):
```
Comment (pereman2) [inline src/pybind/mgr/cli_api/module.py:59]:
```suggestion
            for job_id in range(self._number_of_total_calls):
```

**Review by pereman2 — APPROVED** [review]
nice. I left a minor suggestion

File: src/pybind/mgr/cli_api/module.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        self._number_of_total_calls = number_of_total_calls
+        self._threads = []
+        self._jobs: Queue = Queue()
+        self._time = 0.0
+        self._self_time = []
+        self._lock = threading.Lock()
+
+    def start(self, fn):
+        if(self._number_of_total_calls and self._number_of_parallel_calls):
+            for threadID in range(self._number_of_parallel_calls):
```
Comment (epuertat) [inline src/pybind/mgr/cli_api/module.py:56]: Yup, as @pereman2 suggested, in Python variables are named using the `snake_case` (vs. the `CamelCase` in Typescript for example).
```suggestion
            for thread_id in range(self._number_of_parallel_calls):
```

BTW, [Pylint catches this](http://pylint.pycqa.org/en/latest/user_guide/options.html#predefined-naming-styles). You should add pylint to the .tox where you are running this from. You can copy the pylint settings from the Dashboard .tox.

**Review by epuertat — APPROVED** [review]
LGTM! Please change the snake_case comments from @pereman2 and myself. Take the suggestion of adding pylint as purely optional (but it'd help you catch these issues). Thank you. You're almost there, @Waadkh7 !

File: src/pybind/mgr/cli_api/module.py (hunk 8 above)
Comment (alfonsomthd) [reply src/pybind/mgr/cli_api/module.py]: @sebastian-philipp What do you think about this for `src/pybind/mgr/tox.ini ?
This new check would impact other modules and maybe should be done in a different PR (not cache-related).

Comment (sebastian-philipp) [reply src/pybind/mgr/cli_api/module.py]: Right, you probably want to add `cli_api` to 

https://github.com/ceph/ceph/blob/91afa17c7309b85b1644bf24e769fd403cf2b22b/src/pybind/mgr/tox.ini#L134-L154

Comment (alfonsomthd) [reply src/pybind/mgr/cli_api/module.py]: @sebastian-philipp Sorry I didn't explain myself. @Waadkh7 already did it, my question was about pylint additional check that @epuertat has mentioned.

Comment (sebastian-philipp) [reply src/pybind/mgr/cli_api/module.py]: uh, well. Yeah! Let's add pylint!

File: src/pybind/mgr/tox.ini [hunk 9]
```diff
@@ -1,13 +1,45 @@
...
+[rstlint]
+dirs = README.rst HACKING.rst
+
+[base-pylint]
+commands =
+    pylint {[pylint]addopts} {[pylint]dirs}
+
+[base-lint]
+deps =
+    -rrequirements-lint.txt
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:37]: there's no requirements-lint.txt file

File: src/pybind/mgr/tox.ini [hunk 10]
```diff
@@ -1,13 +1,45 @@
...
+jobs = 1
+dirs = . api controllers plugins services tests ../../../qa/tasks/mgr/dashboard/>
+addopts = -rn --rcfile=.pylintrc --jobs={[pylint]jobs}
+
+[rstlint]
+dirs = README.rst HACKING.rst
+
+[base-pylint]
+commands =
+    pylint {[pylint]addopts} {[pylint]dirs}
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:33]: This overlaps with [testenv:pylint]

File: src/pybind/mgr/tox.ini [hunk 11]
```diff
@@ -1,13 +1,45 @@
...
+wrap_length = 80
+
+[pylint]
+# Allow similarity/code duplication detection
+jobs = 1
+dirs = . api controllers plugins services tests ../../../qa/tasks/mgr/dashboard/>
+addopts = -rn --rcfile=.pylintrc --jobs={[pylint]jobs}
+
+[rstlint]
+dirs = README.rst HACKING.rst
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:29]: These files don't exist for this dir.

File: src/pybind/mgr/tox.ini [hunk 12]
```diff
@@ -1,13 +1,45 @@
...
 
+[base]
+deps =
+    -rrequirements.txt
+
+[isort]
+atomic = true
+multi_line_output = 2
+line_length = 100
+wrap_length = 80
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:20]: Unless you're using isort, this shouldn't be here.

File: src/pybind/mgr/tox.ini [hunk 13]
```diff
@@ -1,13 +1,45 @@
...
+
+[pylint]
+# Allow similarity/code duplication detection
+jobs = 1
+dirs = . api controllers plugins services tests ../../../qa/tasks/mgr/dashboard/>
+addopts = -rn --rcfile=.pylintrc --jobs={[pylint]jobs}
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:26]: These paths only apply to the dashboard.

**Review by epuertat — CHANGES_REQUESTED** [review]
Regarding the pylint changes, you need to do clean-up on the newly added code.

File: src/pybind/mgr/tox.ini [hunk 14]
```diff
@@ -1,13 +1,27 @@
...
+    -rrequirements.txt
+
+[pylint]
+# Allow similarity/code duplication detection
+jobs = 1
+addopts = -rn --rcfile=.pylintrc --jobs={[pylint]jobs}
+
+[base-rst]
+commands =
+    rstcheck --report info --debug -- {[rstlint]dirs}
```
Comment (epuertat) [inline src/pybind/mgr/tox.ini:23]: You can remove this. No .RST file to lint here, right?

**Review by epuertat — APPROVED** [review]
LGMT @Waadkh7 ! But please remove the rst reference in the tox ini.

**Pull request merged**